src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+            // 
+            // public int M_f(int x, int y)
+            // {
+            //     return x * y;
+            // }
+            // 
+            // public void M(int x, int y)
+            // {
+            //     int f = x * y;
+            // }
+            //
+            // public void M1()
+            // {
+            //     M(5, 6, M_f(5, 6)); // This is the generated invocation which is a new argument at the call site
+            // }
+            static TInvocationExpressionSyntax GenerateNewInvocationExpressionForTrampoline(ISyntaxFactsService syntaxFacts,
+                SyntaxEditor editor, SyntaxNode currentInvocation, TInvocationExpressionSyntax invocationExpression,
+                string newMethodIdentifier, int insertionIndex)
+            {
+                var invocationArguments = syntaxFacts.GetArgumentsOfInvocationExpression(currentInvocation);
+                var methodName = editor.Generator.IdentifierName(newMethodIdentifier);
+                var expressionFromInvocation = syntaxFacts.GetExpressionOfInvocationExpression(invocationExpression);
+                var newMethodInvocation = editor.Generator.InvocationExpression(methodName, invocationArguments);
+                var allArguments = invocationArguments.Insert(insertionIndex, newMethodInvocation);
+                return (TInvocationExpressionSyntax)editor.Generator.InvocationExpression(expressionFromInvocation, allArguments);
+            }
+        }
+
+        /// <summary>
+        /// Goes through all of the invocations and replaces the expression with identifiers from the invocation 
+        /// arguments and rewrites the call site with the updated expression as a new argument.
+        /// </summary>
+        private static async Task<Solution> RewriteSolutionWithNewMethodAsync(
+            Document originalDocument, TExpressionSyntax expression, IMethodSymbol methodSymbol,
+            SyntaxNode containingMethod, bool allOccurrences, string parameterName,
+            ImmutableDictionary<Document, ImmutableArray<TInvocationExpressionSyntax>> callSites, bool trampoline,
+            CancellationToken cancellationToken)
+        {
+            var modifiedSolution = originalDocument.Project.Solution;
+
+            foreach (var (project, projectCallSites) in callSites.GroupBy(kvp => kvp.Key.Project))
+            {
+                var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+                foreach (var (document, invocationExpressionList) in projectCallSites)
+                {
+                    modifiedSolution = await ModifyDocumentInvocationsTrampolineAndIntroduceParameterAsync(compilation,
+                        modifiedSolution, document, originalDocument, invocationExpressionList, methodSymbol, containingMethod,
+                        allOccurrences, parameterName, expression, trampoline, cancellationToken).ConfigureAwait(false);
```

Comment:
i don't see why you need to pass in modifiedSolution into this helper.  you're only modifying this document, so just have the documetn return the final syntax root it edited, then have this caller update the modifiedSOlution.  No need to leak that data around.

Reply:
changed